ts-client/src/amm/index.ts

```diff
@@ -0,0 +1,932 @@
+        pool: this.address,
+        user: owner,
+        userAToken,
+        userBToken,
+        aVaultLp: this.poolState.aVaultLp,
+        bVaultLp: this.poolState.bVaultLp,
+        aVaultLpMint: this.vaultA.vaultState.lpMint,
+        bVaultLpMint: this.vaultB.vaultState.lpMint,
+        lpMint: this.poolState.lpMint,
+        tokenProgram: TOKEN_PROGRAM_ID,
+        vaultProgram: this.vaultProgram.programId,
+        userPoolLp,
+      })
+      .remainingAccounts(getRemainingAccounts(this.poolState))
+      .preInstructions(preInstructions)
+      .postInstructions(postInstructions)
+      .transaction();
+
+    return new Transaction({
+      feePayer: owner,
+      ...(await this.program.provider.connection.getLatestBlockhash('finalized')),
+    }).add(depositTx);
+  }
+
+  /**
+   * `getWithdrawQuote` is a function that takes in a withdraw amount and returns the amount of tokens
+   * that will be withdrawn from the pool
+   * @param {BN} withdrawTokenAmount - The amount of tokens you want to withdraw from the pool.
+   * @param {PublicKey} [tokenMint] - The token you want to withdraw. If you want balanced withdraw, leave this blank.
+   * @param {number} [slippage] - The amount of slippage you're willing to accept.
+   * @returns The return value is a tuple of the poolTokenAmountIn, tokenAOutAmount, and
+   * tokenBOutAmount.
+   */
+  public async getWithdrawQuote(
+    withdrawTokenAmount: BN,
+    tokenMint?: PublicKey,
+    slippage?: number,
+  ): Promise<{
+    poolTokenAmountIn: BN;
+    tokenAOutAmount: BN;
+    tokenBOutAmount: BN;
+  }> {
+    const slippageRate = slippage ?? DEFAULT_SLIPPAGE;
+    const { tokenAAmount, tokenBAmount } = await this.getPoolInfo();
+    const { poolLpSupply, vaultALpSupply, vaultBLpSupply, poolVaultALp, poolVaultBLp } = await this.getSplInfo();
+
+    const vaultAWithdrawableAmount = await this.vaultA.getWithdrawableAmount();
+    const vaultBWithdrawableAmount = await this.vaultB.getWithdrawableAmount();
+
+    // balance withdraw
+    if (!tokenMint) {
+      const vaultALpBurn = this.getShareByAmount(withdrawTokenAmount, poolLpSupply, poolVaultALp);
+      const vaultBLpBurn = this.getShareByAmount(withdrawTokenAmount, poolLpSupply, poolVaultBLp);
+
+      const tokenAOutAmount = this.getAmountByShare(vaultALpBurn, vaultAWithdrawableAmount, vaultALpSupply);
+      const tokenBOutAmount = this.getAmountByShare(vaultBLpBurn, vaultBWithdrawableAmount, vaultBLpSupply);
+
+      return {
+        poolTokenAmountIn: withdrawTokenAmount,
+        tokenAOutAmount: getMinAmountWithSlippage(tokenAOutAmount, slippageRate),
+        tokenBOutAmount: getMinAmountWithSlippage(tokenBOutAmount, slippageRate),
+      };
+    }
+
+    // Imbalance withdraw
+    const isWithdrawingTokenA = tokenMint.equals(new PublicKey(this.tokenA.address));
+    const isWithdrawingTokenB = tokenMint.equals(new PublicKey(this.tokenB.address));
+    invariant(isWithdrawingTokenA || isWithdrawingTokenB, ERROR.INVALID_MINT);
```

Comment:
what is these 3 lines checking for?

Reply:
to check` desiredOutputMint` is one of the token of pool.